feat(console): Create backend endpoint for updating a section

aqibmohammadkhan opened the pull request:
## Issue

https://gravitee.atlassian.net/browse/APIM-11947

## Description

- created endpoint and its use case that updates attributes of a navigation item
- It can update title for a  PAGE, LINK and FOLDER
- it can update url or title or both for LINK

`{{baseURL}}/portal-navigation-items/{{navigationId}}`
 
sample payload for update page with required fields

```
{
    "title": "Update title",
    "type": "PAGE",
    "order": 4
}
```


https://github.com/user-attachments/assets/fcc7b2b2-c986-4edd-bf3c-6b753da946a5

<img width="974" height="692" alt="image" src="https://github.com/user-attachments/assets/b04e49a0-2fb2-4b8c-8333-ad72be89024f" />

<img width="990" height="741" alt="image" src="https://github.com/user-attachments/assets/02ab1bd8-3d88-453d-a8f1-39413ac4cb38" />



Response

```
{
    "type": "LINK",
    "url": "yahoo.com",
    "id": "70dcfdcd-0b83-43fb-aa9f-7d948295742c",
    "organizationId": "DEFAULT",
    "environmentId": "DEFAULT",
    "title": "this is updated title",
    "area": "TOP_NAVBAR",
    "order": 4
}
```

## Additional context

<!-- Add any other context about the PR here -->
<!-- It can be links to other PRs or docs or drawing -->
<!-- Or reproduction steps in case of bug fix -->

aqibmohammadkhan committed:
feat(console): Create backend endpoint for updating a section

jourdiw commented on gravitee-apim-rest-api/gravitee-apim-rest-api-service/src/main/java/io/gravitee/apim/core/portal_page/use_case/UpdatePortalNavigationItemUseCase.java:
```diff
@@ -0,0 +1,48 @@
+        var navItem = input.updatePortalNavigationItem;
+        PortalNavigationItem item = portalNavigationItemsQueryService.findByIdAndEnvironmentId(input.environmentId(), navItem.getId());
+        if (item == null) {
+            throw new InvalidDataException("Portal navigation item [%s] not found.".formatted(navItem.getId()));
```
What do you think of this being a 404 instead of a 400 error?

aqibmohammadkhan replied:
Done. changed to ResourceNotFoundException.

jourdiw commented on gravitee-apim-rest-api/gravitee-apim-rest-api-service/src/main/java/io/gravitee/apim/core/portal_page/use_case/UpdatePortalNavigationItemUseCase.java:
You can add the URL validation that is used in the create 👍

aqibmohammadkhan replied:
done. also added additional validations like type consistency.

jourdiw reviewed (commented):
Personally I try to avoid mocking the use cases for the resource tests. One of the great advantages with the Onion architecture is that we can have better confidence for the end to end functionality. 

WDYT about not mocking the use cases for the resources?

kovaren commented on gravitee-apim-rest-api/gravitee-apim-rest-api-management-v2/gravitee-apim-rest-api-management-v2-rest/src/main/java/io/gravitee/rest/api/management/v2/rest/resource/environment/PortalNavigationItemsResource.java:
```diff
@@ -54,6 +56,9 @@ public class PortalNavigationItemsResource extends AbstractResource {
     private ListPortalNavigationItemsUseCase listPortalNavigationItemsUseCase;
 
+    @Inject
+    private UpdatePortalNavigationItemUseCase updatePortalNavigationItemUseCase;
```
Based on the pattern used by our existing resources (e.g. SharedPolicyGroupsResource), the update use case should be part of PortalNavigationItemResource and that resource should be nested under PortalNavigationItemsResource using a path parameter.

aqibmohammadkhan replied:
Done. created a new sub resource.

kovaren commented on gravitee-apim-rest-api/gravitee-apim-rest-api-management-v2/gravitee-apim-rest-api-management-v2-rest/src/main/java/io/gravitee/rest/api/management/v2/rest/resource/environment/PortalNavigationItemsResource.java:
```diff
@@ -93,4 +98,26 @@ public Response createPortalNavigationItem(@Valid @NotNull final BaseCreatePorta
+    @Produces(MediaType.APPLICATION_JSON)
+    @Permissions({ @Permission(value = RolePermission.ENVIRONMENT_DOCUMENTATION, acls = RolePermissionAction.UPDATE) })
+    public Response updatePortalNavigationItem(
+        @QueryParam("navId") String navigationItemId,
```
Related to https://github.com/gravitee-io/gravitee-api-management/pull/14143/files#r2548981079, `navId` should be a \@PathParam

aqibmohammadkhan replied:
done.

aqibmohammadkhan commented:
> Personally I try to avoid mocking the use cases for the resource tests. One of the great advantages with the Onion architecture is that we can have better confidence for the end to end functionality.
> 
> WDYT about not mocking the use cases for the resources?

I agree. This was already done, so i just added my changes below that. I have removed all the new created use cases now and mocking the in-memory crud and other services.

kovaren commented on gravitee-apim-rest-api/gravitee-apim-rest-api-management-v2/gravitee-apim-rest-api-management-v2-rest/src/main/resources/openapi/openapi-environments.yaml:
```diff
@@ -1871,6 +1901,81 @@ components:
+        parentId:
+          type: string
+          format: uuid
+          description: The parent ID of the navigation item, if not provided item is created at root
```
```suggestion
          description: The parent ID of the navigation item, if not provided item is moved to root level

kovaren commented on gravitee-apim-rest-api/gravitee-apim-rest-api-management-v2/gravitee-apim-rest-api-management-v2-rest/src/test/java/io/gravitee/rest/api/management/v2/rest/resource/environment/PortalNavigationItemResource_PutTest.java:
```diff
@@ -0,0 +1,226 @@
+
+    @Test
+    void should_update_portal_navigation_item_title_link() {
+        // Given an existing PAGE item id from fixtures
```
```suggestion
        // Given an existing LINK item id from fixtures

kovaren commented on gravitee-apim-rest-api/gravitee-apim-rest-api-management-v2/gravitee-apim-rest-api-management-v2-rest/src/test/java/io/gravitee/rest/api/management/v2/rest/resource/environment/PortalNavigationItemResource_PutTest.java:
```diff
@@ -0,0 +1,226 @@
+
+    @Test
+    void should_update_portal_navigation_item_title_folder() {
+        // Given an existing PAGE item id from fixtures
```
```suggestion
        // Given an existing FOLDER item id from fixtures

kovaren commented on gravitee-apim-rest-api/gravitee-apim-rest-api-service/src/main/java/io/gravitee/apim/core/portal_page/model/PortalNavigationFolder.java:
```diff
@@ -29,4 +30,12 @@ public PortalNavigationFolder(
+        super.update(navItem);
+        if (this.getTitle().isBlank()) {
+            throw new InvalidDataException("Title is required for FOLDER navigation items.");
+        }
```
1. I would say that all validation should be done in one place, for example the validator service.

2. I agree that the blank value is invalid but regarding the null value, it would just mean that we keep the current value. Let me know if that works for you.

kovaren commented on gravitee-apim-rest-api/gravitee-apim-rest-api-service/src/main/java/io/gravitee/apim/core/portal_page/domain_service/CreatePortalNavigationItemValidatorService.java:
Since this validator is now also used for the update, I would suggest renaming it to smth like:
```suggestion
public class PortalNavigationItemValidatorService {

kovaren commented on gravitee-apim-rest-api/gravitee-apim-rest-api-management-v2/gravitee-apim-rest-api-management-v2-rest/src/main/resources/openapi/openapi-environments.yaml:
```diff
@@ -1871,6 +1901,81 @@ components:
           FOLDER: "#/components/schemas/CreatePortalNavigationFolder"
           LINK: "#/components/schemas/CreatePortalNavigationLink"
 
+    BaseUpdatePortalNavigationItem:
```
As discussed with the team, `id`, `type` and `area` don't need to be on BaseUpdatePortalNavigationItem because they are not updatable.

kovaren commented on gravitee-apim-rest-api/gravitee-apim-rest-api-management-v2/gravitee-apim-rest-api-management-v2-rest/src/main/resources/openapi/openapi-environments.yaml:
```diff
@@ -1871,6 +1901,81 @@ components:
+          FOLDER: "#/components/schemas/UpdatePortalNavigationFolder"
+          PAGE: "#/components/schemas/UpdatePortalNavigationPage"
+          LINK: "#/components/schemas/UpdatePortalNavigationLink"
+      required: [ title, type, area ]
```
I would argue that no properties are required here or on child schemas. If a property is null, we keep the current value. This would allow to only specify what we want to update.
WDYT?

jourdiw replied:
If it's a POST for the update, I agree that we can let fields be null. However if it's a PATCH, I would expect the data sent to be exactly what's put into the object in the DB. 🤔

aqibmohammadkhan replied:
as discussed, keeping it as it is.

jourdiw commented on gravitee-apim-rest-api/gravitee-apim-rest-api-management-v2/gravitee-apim-rest-api-management-v2-rest/src/main/java/io/gravitee/rest/api/management/v2/rest/resource/environment/PortalNavigationItemResource.java:
```diff
@@ -0,0 +1,65 @@
+    @Permissions({ @Permission(value = RolePermission.ENVIRONMENT_DOCUMENTATION, acls = RolePermissionAction.UPDATE) })
+    public Response updatePortalNavigationItem(
+        @PathParam("navId") String navigationItemId,
+        @Valid @NotNull final BaseUpdatePortalNavigationItem updatePortalNavigationItem
```
You can use your `UpdatePortalNavigationItem` here 👍 
This will help handle the polymorphism

kovaren replied:
Just my 2 cents, when I tried to use CreatePortalNavigationItem instead of BaseCreatePortalNavigationItem for the creation request, it broke the required props validation.

It might be less relevant for the update since none of the props are required for now at least but I would still stick with the BaseUpdatePortalNavigationItem for the sake of consistency.

But I leave it to you @aqibmohammadkhan

jourdiw replied:
Hmmm in theory we got it working with the `ApiPagesResource` --> https://github.com/gravitee-io/gravitee-api-management/blob/master/gravitee-apim-rest-api/gravitee-apim-rest-api-management-v2/gravitee-apim-rest-api-management-v2-rest/src/main/java/io/gravitee/rest/api/management/v2/rest/resource/documentation/ApiPagesResource.java#L153

kovaren replied:
That's because UpdateDocumentation is the equivalent of BaseUpdatePortalNavigationItem, it has all the common fields inherited by the children, it's just different naming.
UpdatePortalNavigationItem however is the composition type, which is needed by generators to correctly generate client libraries

aqibmohammadkhan replied:
As discussed, keeping it as it is.

kovaren commented on gravitee-apim-rest-api/gravitee-apim-rest-api-automation/gravitee-apim-rest-api-automation-rest/src/test/java/io/gravitee/apim/rest/api/automation/spring/ResourceContextConfiguration.java:
```diff
@@ -867,8 +872,27 @@ public CreatePortalNavigationItemUseCase createPortalNavigationItemUseCase() {
+        PortalNavigationItemsQueryService portalNavigationItemsQueryService,
+        PortalNavigationItemValidatorService validatorService
+    ) {
+        return new UpdatePortalNavigationItemUseCase(portalNavigationItemCrudService, portalNavigationItemsQueryService, validatorService);
```
Is this not being mocked related to [this suggestion](https://github.com/gravitee-io/gravitee-api-management/pull/14143#pullrequestreview-3491661357)?

It's kind of breaking the pattern that we have for all the other use cases, not sure we want to do it in this particular PR, maybe just confirm with @jourdiw.

If it's what we want to do, i'm fine with it, you can resolve this comment

aqibmohammadkhan replied:
this is returning the new use case with the in memory services and not the mock. 
mocking the use case, which is return mock(use case) is something we should not do.

kovaren replied:
Ok I double checked, we already have non mocked services so it's not an issue with breaking the pattern, let's keep it.

kovaren commented on gravitee-apim-rest-api/gravitee-apim-rest-api-management-v2/gravitee-apim-rest-api-management-v2-rest/src/main/resources/openapi/openapi-environments.yaml:
```diff
@@ -1871,6 +1905,68 @@ components:
+        order:
+          type: integer
+          minimum: 0
+          description: The order of the navigation item, from 0 to MAX + 1, where MAX is the maximum existing order within the target parent. If not provided or greater than MAX, the new item is appended to the end of the target parent.
```
I missed this on the last review
```suggestion
          description: The order of the navigation item, from 0 to MAX, where MAX is the maximum existing order within the target parent. If greater than MAX, the item moved to the end of the target parent.

kovaren commented on gravitee-apim-rest-api/gravitee-apim-rest-api-service/src/main/java/io/gravitee/apim/core/portal_page/domain_service/PortalNavigationItemValidatorService.java:
```diff
@@ -106,4 +111,36 @@ private boolean isValidUrl(String url) {
+    public void validateToUpdate(UpdatePortalNavigationItem toUpdate, PortalNavigationItem existingItem) {
+        enforceTypeConsistency(existingItem, toUpdate.getType());
+
+        if (toUpdate.getTitle() == null || toUpdate.getTitle().isBlank()) {
```
Null is handled at the spec level
```suggestion
        if (toUpdate.getTitle().isBlank()) {

kovaren commented on gravitee-apim-rest-api/gravitee-apim-rest-api-service/src/main/java/io/gravitee/apim/core/portal_page/model/UpdatePortalNavigationItem.java:
```diff
@@ -0,0 +1,37 @@
+
+    private PortalNavigationItemId id;
+    private String title;
+    private PortalArea area;
```
This object shouldn't have `id` and `area` in accordance with the open api spec.
```suggestion
    private String title;

kovaren commented on gravitee-apim-rest-api/gravitee-apim-rest-api-service/src/main/java/io/gravitee/apim/core/portal_page/domain_service/PortalNavigationItemValidatorService.java:
```diff
@@ -106,4 +111,36 @@ private boolean isValidUrl(String url) {
+            String url = toUpdate.getUrl();
+            if (url == null || url.isBlank()) {
+                throw new InvalidDataException("URL is required for LINK navigation items.");
+            }
```
Null is handled by the spec and the blank value by the `isValidUrl()` method
```suggestion

jourdiw reviewed (approved):
Nice job!

A couple things if you are feeling 🚀 :
- Use core-based exceptions in the services rather than the `InvalidDataException`
- Removing page content id from the Update model in the core package to match what is expected from the Resource 👍
